File: zy73133/tide_normalizer.py
from __future__ import annotations

import re
from typing import Tuple, Optional, List

from models import NormalizedTide, TideUnit


UNIT_ALIASES = {
    TideUnit.METERS: ['m', 'meter', 'meters', 'metre', 'metres', '米', '公尺'],
    TideUnit.CENTIMETERS: ['cm', 'centimeter', 'centimeters', 'centimetre', 'centimetres', '厘米', '公分'],
    TideUnit.FEET: ['ft', 'feet', 'foot', '英尺', '呎'],
}

FEET_TO_METERS = 0.3048
CM_TO_METERS = 0.01


def _detect_unit(raw_unit: str) -> Tuple[TideUnit, str]:
    if raw_unit is None:
        return TideUnit.UNKNOWN, ''

    cleaned = raw_unit.strip().lower()
    if not cleaned:
        return TideUnit.UNKNOWN, raw_unit

    for unit, aliases in UNIT_ALIASES.items():
        if cleaned in [a.lower() for a in aliases]:
            return unit, raw_unit

    for unit, aliases in UNIT_ALIASES.items():
        for alias in aliases:
            if alias.lower() in cleaned:
                return unit, raw_unit

    return TideUnit.UNKNOWN, raw_unit


def _extract_value_and_unit(raw_value: str, raw_unit: str) -> Tuple[Optional[float], TideUnit, str, List[str]]:
    notes: List[str] = []
    original_value_raw = raw_value.strip() if raw_value else ''
    combined = f"{raw_value or ''} {raw_unit or ''}".strip()

    if raw_unit:
        detected_unit, unit_raw = _detect_unit(raw_unit)
    else:
        detected_unit = TideUnit.UNKNOWN
        unit_raw = ''

    value_match = re.search(r'(-?\d+\.?\d*)', original_value_raw)
    if not value_match:
        value_match = re.search(r'(-?\d+\.?\d*)', combined)

    if not value_match:
        return None, detected_unit, unit_raw, [f"无法从 '{combined}' 中提取数值"]

    try:
        raw_numeric = float(value_match.group(1))
    except ValueError:
        return None, detected_unit, unit_raw, [f"数值解析失败: {value_match.group(1)}"]

    if detected_unit == TideUnit.UNKNOWN:
        unit_from_combined = re.search(r'(m|cm|ft|米|厘米|公分|英尺|呎)', combined, re.IGNORECASE)
        if unit_from_combined:
            detected_unit, _ = _detect_unit(unit_from_combined.group(1))
            if detected_unit != TideUnit.UNKNOWN:
                notes.append(f"从数值串中识别单位: {unit_from_combined.group(1)}")
                unit_raw = unit_from_combined.group(1)

    return raw_numeric, detected_unit, unit_raw if unit_raw else raw_unit, notes


def normalize_tide(raw_value: str, raw_unit: str) -> NormalizedTide:
    raw_numeric, detected_unit, original_unit_raw, notes = _extract_value_and_unit(raw_value, raw_unit)

    if raw_numeric is None:
        return NormalizedTide(
            value_meters=0.0,
            original_value=raw_value or '',
            original_unit=TideUnit.UNKNOWN,
            original_unit_raw=raw_unit or '',
            normalize_notes=notes
        )

    if detected_unit == TideUnit.METERS:
        value_m = raw_numeric
    elif detected_unit == TideUnit.CENTIMETERS:
        value_m = raw_numeric * CM_TO_METERS
        notes.append(f"厘米转米: {raw_numeric} cm × {CM_TO_METERS} = {value_m} m")
    elif detected_unit == TideUnit.FEET:
        value_m = raw_numeric * FEET_TO_METERS
        notes.append(f"英尺转米: {raw_numeric} ft × {FEET_TO_METERS} = {value_m:.4f} m")
    else:
        value_m = raw_numeric
        notes.append(
            f"单位未识别，保留原始数值 {raw_numeric}，按米处理（需人工复核）。原始单位字符串: '{raw_unit or '(空)'}"
        )

    return NormalizedTide(
        value_meters=round(value_m, 6),
        original_value=raw_value or '',
        original_unit=detected_unit,
        original_unit_raw=original_unit_raw or (raw_unit or ''),
        normalize_notes=notes
    )


def tide_to_original_string(tide: NormalizedTide) -> str:
    if tide.original_unit == TideUnit.UNKNOWN:
        return f"{tide.original_value} (单位未知)"
    unit_display = {
        TideUnit.METERS: 'm',
        TideUnit.CENTIMETERS: 'cm',
        TideUnit.FEET: 'ft',
    }.get(tide.original_unit, tide.original_unit_raw)
    return f"{tide.original_value} {unit_display}".strip()

File: zy73133/src/engine/tideNormalizer.ts
import type { NormalizedTide, TideUnit } from './types';

const FEET_TO_METERS = 0.3048;
const CM_TO_METERS = 0.01;

const UNIT_ALIASES: Record<TideUnit, string[]> = {
  m: ['m', 'meter', 'meters', 'metre', 'metres', '米', '公尺'],
  cm: ['cm', 'centimeter', 'centimeters', 'centimetre', 'centimetres', '厘米', '公分'],
  ft: ['ft', 'feet', 'foot', '英尺', '呎'],
  unknown: [],
};

function detectUnit(rawUnit: string): { unit: TideUnit; raw: string } {
  if (rawUnit == null) return { unit: 'unknown', raw: '' };
  const cleaned = String(rawUnit).trim().toLowerCase();
  if (!cleaned) return { unit: 'unknown', raw: rawUnit };

  for (const unit of Object.keys(UNIT_ALIASES) as TideUnit[]) {
    if (UNIT_ALIASES[unit].some((a) => a.toLowerCase() === cleaned)) {
      return { unit, raw: rawUnit };
    }
  }
  for (const unit of Object.keys(UNIT_ALIASES) as TideUnit[]) {
    for (const alias of UNIT_ALIASES[unit]) {
      if (alias.toLowerCase() && cleaned.includes(alias.toLowerCase())) {
        return { unit, raw: rawUnit };
      }
    }
  }
  return { unit: 'unknown', raw: rawUnit };
}

interface Extracted {
  value: number | null;
  unit: TideUnit;
  unitRaw: string;
  notes: string[];
}

function extractValueAndUnit(rawValue: string, rawUnit: string): Extracted {
  const notes: string[] = [];
  const originalValueRaw = (rawValue ?? '').trim();
  const combined = `${rawValue ?? ''} ${rawUnit ?? ''}`.trim();

  let detectedUnit: TideUnit = 'unknown';
  let unitRaw = '';
  if (rawUnit) {
    const det = detectUnit(rawUnit);
    detectedUnit = det.unit;
    unitRaw = det.raw;
  }

  let valueMatch = originalValueRaw.match(/(-?\d+\.?\d*)/);
  if (!valueMatch) valueMatch = combined.match(/(-?\d+\.?\d*)/);

  if (!valueMatch) {
    return { value: null, unit: detectedUnit, unitRaw, notes: [`无法从 '${combined}' 中提取数值`] };
  }

  const rawNumeric = parseFloat(valueMatch[1]);
  if (Number.isNaN(rawNumeric)) {
    return { value: null, unit: detectedUnit, unitRaw, notes: [`数值解析失败: ${valueMatch[1]}`] };
  }

  if (detectedUnit === 'unknown') {
    const unitFromCombined = combined.match(/(m|cm|ft|米|厘米|公分|英尺|呎)/i);
    if (unitFromCombined) {
      const det = detectUnit(unitFromCombined[1]);
      if (det.unit !== 'unknown') {
        detectedUnit = det.unit;
        notes.push(`从数值串中识别单位: ${unitFromCombined[1]}`);
        unitRaw = unitFromCombined[1];
      }
    }
  }

  return {
    value: rawNumeric,
    unit: detectedUnit,
    unitRaw: unitRaw || rawUnit || '',
    notes,
  };
}

export function normalizeTide(rawValue: string, rawUnit: string): NormalizedTide {
  const { value: rawNumeric, unit: detectedUnit, unitRaw, notes } = extractValueAndUnit(rawValue, rawUnit);

  if (rawNumeric == null) {
    return {
      valueMeters: null,
      originalValue: rawValue ?? '',
      originalUnit: 'unknown',
      originalUnitRaw: rawUnit ?? '',
      normalizeNotes: notes,
    };
  }

  let valueM = rawNumeric;
  if (detectedUnit === 'm') {
    // keep as meters
  } else if (detectedUnit === 'cm') {
    valueM = rawNumeric * CM_TO_METERS;
    notes.push(`厘米转米: ${rawNumeric} cm × ${CM_TO_METERS} = ${valueM} m`);
  } else if (detectedUnit === 'ft') {
    valueM = rawNumeric * FEET_TO_METERS;
    notes.push(`英尺转米: ${rawNumeric} ft × ${FEET_TO_METERS} = ${valueM.toFixed(4)} m`);
  } else {
    notes.push(
      `单位未识别，保留原始数值 ${rawNumeric}，按米处理（需人工复核）。原始单位字符串: '${rawUnit || '(空)'}'`,
    );
  }

  return {
    valueMeters: Math.round(valueM * 1e6) / 1e6,
    originalValue: rawValue ?? '',
    originalUnit: detectedUnit,
    originalUnitRaw: unitRaw || rawUnit || '',
    normalizeNotes: notes,
  };
}

export function tideToOriginalString(tide: NormalizedTide | null): string {
  if (!tide) return 'N/A';
  if (tide.originalUnit === 'unknown') {
    return tide.originalValue ? `${tide.originalValue} (单位未知)` : 'N/A';
  }
  const unitDisplay: Record<TideUnit, string> = { m: 'm', cm: 'cm', ft: 'ft', unknown: tide.originalUnitRaw };
  return `${tide.originalValue} ${unitDisplay[tide.originalUnit]}`.trim();
}
